### ISSUE TITLE
FuncTickFormatter broken with Categorical Axis

### ISSUE THREAD

**Issue by hiramf** [bokeh/bokeh#7015]
Bokeh v. 0.12.10dev2

In attempting to customize axis tick labels, there are two ways to achieve this (python function and javascript).
```python
from bokeh.models FuncTickFormatter
from bokeh.plotting import figure, output_notebook
output_notebook()
p = figure(plot_width=500, plot_height=500)
p.circle([0, 2, 4, 6, 8, 10], [6, 2, 4, 10, 8, 0], size=30)
p.yaxis.formatter = FuncTickFormatter(code="""
    return 'foo'
""")
def ticker():
    return "bar"
p.xaxis.formatter = FuncTickFormatter.from_py_func(ticker)
show(p)
```

However, attempting to do so on a categorical plot does not work (on the categorical axis):
```python
from bokeh.io import show, output_file
from bokeh.models import ColumnDataSource, FuncTickFormatter
from bokeh.palettes import Spectral5
from bokeh.plotting import figure
from bokeh.sampledata.autompg import autompg as df
from bokeh.transform import factor_cmap

df.cyl = df.cyl.astype(str)
group = df.groupby('cyl')

source = ColumnDataSource(group)

cyl_cmap = factor_cmap('cyl', palette=Spectral5, factors=sorted(df.cyl.unique()))

p = figure(plot_height=350, x_range=group, title="MPG by # Cylinders",
           toolbar_location=None, tools="")

p.vbar(x='cyl', top='mpg_mean', width=1, source=source,
       line_color=cyl_cmap, fill_color=cyl_cmap)

p.y_range.start = 0
p.xgrid.grid_line_color = None
p.xaxis.axis_label = "some stuff"
p.xaxis.major_label_orientation = 1.2
p.outline_line_color = None

# TICKER FUNCTIONS

def ticker():
    return "bar"
# CATEGORICAL TICKS SHOULD ALL BE  'bar' (BROKEN)
p.xaxis.formatter = FuncTickFormatter.from_py_func(ticker) # CATEGORICAL AXIS

# LINEAR TICKS ARE ALL 'foo' (WORKS)
p.yaxis.formatter = FuncTickFormatter(code="""
    return 'foo'
""")

show(p)
```

**Comment by hafen** [comment]
I'm experiencing the same thing in 0.12.9. It worked in 0.12.6. The callback is not called at all in the CategoricalAxis case.

**Comment by bryevdv** [comment]
I've merged #7025 I don't think it's necessarily a full solution for all cases but it covers what's described in this specific issue, and more work can wait until there is a demand for it and feedback to help figure out the best approach.
